### PR TITLE
feat: add switch 

### PR DESCRIPTION
**Pull request by tanvibhakta**
## Description

Fixes https://github.com/appsmithorg/appsmith/issues/19321

Zeplin: https://zpl.io/Nwd7B59

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
https://github.com/appsmithorg/TestSmith/issues/2141

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/design-system-v2/src/Toggle/Toggle.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+import { ComponentMeta, ComponentStory } from "@storybook/react";
+
+import { Toggle } from "./Toggle";
+
+export default {
+  title: "Design System V2/Toggle",
+  component: Toggle,
+} as ComponentMeta<typeof Toggle>;
+
+export const Basic = <Toggle />;
```
Comment (albinAppsmith) [inline packages/design-system-v2/src/Toggle/Toggle.stories.tsx:11]: Can we make this not nested? I don't see a variation for toggle in future.

File: packages/design-system-v2/src/Toggle/Toggle.styles.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    width: 1rem;
+    height: 1rem;
+    grid-area: off;
+    border-radius: var(--ads-border-radius-circle);
+    background-color: var(--thumb-color-bg);
+  }
+
+  &:hover {
+    // TODO: needs a theme variable
+    --toggle-color-bg: var(--ads-v2-color-black-250)
```
Comment (albinAppsmith) [inline packages/design-system-v2/src/Toggle/Toggle.styles.tsx:39]: we do have theme variable, --ads-v2-color-bg-emphasis

File: packages/design-system-v2/src/Toggle/Toggle.styles.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  &:checked {
+    justify-content: flex-end;
+    --toggle-color-bg: var(--ads-v2-color-bg-brand-secondary);
+
+    // set position of thumb when checked
+    &::before {
+      grid-area: on;
+    }
```
Comment (albinAppsmith) [inline packages/design-system-v2/src/Toggle/Toggle.styles.tsx:54]: :hover state for on state is also needed.

File: packages/design-system-v2/src/Toggle/Toggle.styles.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  &:hover {
+    // TODO: needs a theme variable
+    --toggle-color-bg: var(--ads-v2-color-black-250)
+  }
+
+  &:disabled {
+    cursor: not-allowed;
+    opacity: 0.5;
```
Comment (albinAppsmith) [inline packages/design-system-v2/src/Toggle/Toggle.styles.tsx:44]: opacity is 0.6, may be we can add it in theme

File: packages/design-system-v2/src/Toggle/Toggle.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+import { StyledToggle } from "./Toggle.styles";
+import { ToggleProps } from "./Toggle.types";
+
+const Toggle = (props: ToggleProps) => {
+  const { checked, ...rest } = props;
```
Comment (albinAppsmith) [inline packages/design-system-v2/src/Toggle/Toggle.tsx:6]: onChange needs to be destructured, else we will face issue with main repo

File: packages/design-system-v2/src/Toggle/Toggle.styles.tsx (hunk 2 above)
Comment (tanvibhakta) [reply packages/design-system-v2/src/Toggle/Toggle.styles.tsx]: that is the background color already. We need a theme variable for black-250.

File: packages/design-system-v2/src/Toggle/Toggle.styles.tsx (hunk 3 above)
Comment (tanvibhakta) [reply packages/design-system-v2/src/Toggle/Toggle.styles.tsx]: We apparently only have hover for default state, not active. https://www.figma.com/file/lWkmRog0cbht9upiFJVs4s?node-id=15:4280#347892275

File: .changeset/moody-files-enjoy.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+"design-system-v2": patch
```
Comment (albinAppsmith) [inline .changeset/moody-files-enjoy.md:2]:
```suggestion
```

File: packages/design-system/src/__theme__/default/index.css [hunk 7]
```diff
@@ -1,6 +1,10 @@
 @import "colors.css";
 @import "variables.css";
 
+*:focus-visible {
```
Comment (albinAppsmith) [inline packages/design-system/src/__theme__/default/index.css:4]: Remove this

**Review by albinAppsmith — APPROVED** [review]

File: packages/design-system/src/__theme__/default/index.css [hunk 8]
```diff
@@ -119,6 +121,8 @@
     * ===========================================*
   */
   --ads-v2-border-radius: 4px;
+  --ads-v2-pill-border-radius: 10px;
```
Comment (albinAppsmith) [inline packages/design-system/src/__theme__/default/index.css:124]: Do you see any other use case for this variable? other than switch?

File: packages/design-system/src/Switch/Switch.styles.tsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    css`
+      cursor: not-allowed;
+      opacity: var(--ads-v2-opacity-disabled);
+    `}
+`;
+
+export const StyledSwitchInput = styled.input<{
+  isDisabled?: boolean;
+  isSelected?: boolean;
+  onClick?: any;
```
Comment (albinAppsmith) [inline packages/design-system/src/Switch/Switch.styles.tsx:34]: Why onClick type declared here?

File: packages/design-system/src/__theme__/default/index.css (hunk 8 above)
Comment (tanvibhakta) [reply packages/design-system/src/__theme__/default/index.css:124]: not particularly but wasn't sure about hardcoding it - shall I?

File: packages/design-system/src/Switch/Switch.styles.tsx (hunk 9 above)
Comment (tanvibhakta) [reply packages/design-system/src/Switch/Switch.styles.tsx:34]: Will remove.

File: packages/design-system/src/__theme__/default/index.css (hunk 8 above)
Comment (albinAppsmith) [reply packages/design-system/src/__theme__/default/index.css:124]: Yeah seems like this is only foe switch component. So keep it inside that component only

**Review by albinAppsmith — APPROVED** [review]

**Comment by shadabbuchh** [comment]
Bugs in this PR:

**Functional:**

1.  When we enter a long text in the `children` control, the switch component is not completely visible.
Screenshot 2023-02-06 at 8 13 01 PM

2. Upon selecting a value for `isSelected`, we cannot change the selection from the canvas.
3. In disabled state we get a hover state
https://www.loom.com/share/50af0b9228cb4f9dbbe799d4db5d7f5a

**Design:**

1. Line height for the Label is 19px whereas the expected value is 20 px.
2. Letter spacing is -0.24px, expected is -0.1px.
3. Wrong background colour for unselected -> hover state.
5. Not able to see Border colour in Focus state.
6. Focus state looks different in different browsers: https://www.loom.com/share/f57abf49a3ca48a88305de57b8dc16fc

**Comment by tanvibhakta** [comment]
Functional:

>    When we enter a long text in the children control, the switch component is not completely visible.
Fixed.

>    Upon selecting a value for isSelected, we cannot change the selection from the canvas.
This happens across components for a "controlled" state. Controlled means that the programmer now needs to implement the control if they want the value of it. 

>    In disabled state we get a hover state
This was happening only when the switch was selected and then disabled. Fixed.


Design:

>    Line height for the Label is 19px whereas the expected value is 20 px.
>    Letter spacing is -0.24px, expected is -0.1px.
We're not looking at typography here, right? 

>    Wrong background colour for unselected -> hover state.
fixed

>    Not able to see Border colour in Focus state.
>    Focus state looks different in different browsers: https://www.loom.com/share/f57abf49a3ca48a88305de57b8dc16fc
both should be fixed now.

**Comment by shadabbuchh** [comment]
All of the above issues are fixed.
Additional issues noticed:

1. No hover state when the toggle is selected
2. Wrong Box shadow in all states when the toggle is selected.
@tanvibhakta

**Comment by shadabbuchh** [comment]
All issues are fixed. PR is good to go.

**Pull request merged**